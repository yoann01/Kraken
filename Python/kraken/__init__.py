"""Kraken Framework."""

import logging
import os

__all__ = ['core', 'helpers', 'plugins', 'ui']


krakenPath = os.path.abspath(os.path.join(os.path.dirname(os.path.realpath(__file__)), '..', '..'))
if os.environ.get('KRAKEN_PATH', None) is None:
    os.environ['KRAKEN_PATH'] = krakenPath

krakenExtsPath = os.path.join(krakenPath, 'Exts')
if 'FABRIC_EXTS_PATH' in os.environ:
    if krakenExtsPath not in os.environ['FABRIC_EXTS_PATH']:
        os.environ['FABRIC_EXTS_PATH'] = krakenExtsPath + os.pathsep + os.environ['FABRIC_EXTS_PATH']

krakenPresetsPath = os.path.join(krakenPath, 'Presets', 'DFG')
if 'FABRIC_DFG_PATH' in os.environ:
    if krakenPresetsPath not in os.environ['FABRIC_DFG_PATH']:
        os.environ['FABRIC_DFG_PATH'] = krakenPresetsPath + os.pathsep + os.environ['FABRIC_DFG_PATH']


logging.basicConfig(format='[KRAKEN] %(levelname)s: %(message)s', level=logging.INFO)

# Custom inform level for use with UI label getting added to the status bar.
logging.INFORM = 25
logging.addLevelName(logging.INFORM, 'INFORM')
logging.Logger.inform = lambda inst, msg, *args, **kwargs: inst.log(logging.INFORM, msg, *args, **kwargs)


def release():
    print "RELEASE THE KRAKEN!!!!"
