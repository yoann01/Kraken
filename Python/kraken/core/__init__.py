"""Kraken Core."""

VERSION_MAJOR = 1
VERSION_MINOR = 2
VERSION_BUILD = 6
VERSION_SUFFIX = ""


def getVersion():
    """Contatenates the version globals and returns the current version of
    Kraken.

    Returns:
        str: Current version of Kraken.

    """

    versionString = str(VERSION_MAJOR) + "." + str(VERSION_MINOR) + "." + str(VERSION_BUILD)
    if VERSION_SUFFIX:
        versionString = versionString + "-" + VERSION_SUFFIX

    return versionString
