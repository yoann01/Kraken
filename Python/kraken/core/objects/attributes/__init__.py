"""Kraken Attributes."""

__all__ = ['attribute_group',
           'attribute',
           'bool_attribute',
           'color_attribute',
           'scalar_attribute',
           'integer_attribute',
           'string_attribute']
