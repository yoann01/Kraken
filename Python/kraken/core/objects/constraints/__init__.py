"""Kraken Constraints."""

__all__ = ['constraint',
           'scale_constraint',
           'orientation_constraint',
           'position_constraint',
           'pose_constraint']