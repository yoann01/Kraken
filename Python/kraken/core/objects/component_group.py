"""Kraken - objects.component_group module.

Classes:
ComponentGroup -- Hiearchy group representation.

"""

from kraken.core.objects.object_3d import Object3D


class ComponentGroup(Object3D):
    """ComponentGroup object."""

    def __init__(self, name, component, parent=None, metaData=None):
        super(ComponentGroup, self).__init__(name, parent=parent, metaData=metaData)

        self.setComponent(component)

        self.setShapeVisibility(False)
        self.lockRotation(True, True, True)
        self.lockScale(True, True, True)
        self.lockTranslation(True, True, True)


    # =============
    # Name Methods
    # =============
    def getName(self):
        """Gets the decorated name of the object.

        Returns:
            str: Decorated name of the object.

        """

        # During construction of the base class, the name is tested before
        # the component is assigned.
        if self.getComponent() is not None:
            # The ComponentGroup's name should always match the component's name.
            return self.getComponent().getName()
        else:
            return super(ComponentGroup, self).getName()


    def getNameDecoration(self):
        """Gets the decorated name of the object.

        Returns:
            str: Decorated name of the object.

        """

        if self.getComponent() is not None:
            # The ComponentGroup's name should always match the component's name.
            return self.getComponent().getNameDecoration()
        else:
            return super(ComponentGroup, self).getNameDecoration()
