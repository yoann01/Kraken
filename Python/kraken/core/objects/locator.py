"""Kraken - objects.locator module.

Classes:
Locator -- Locator representation.

"""

from kraken.core.objects.object_3d import Object3D


class Locator(Object3D):
    """Locator object."""

    def __init__(self, name, parent=None, flags=None, metaData=None):
        super(Locator, self).__init__(name, parent=parent, flags=flags, metaData=metaData)
