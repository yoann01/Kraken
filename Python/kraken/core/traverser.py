"""Kraken - traverser module.

Classes:
Traverser - Base Traverser.

"""

from kraken.core.objects.scene_item import SceneItem
from kraken.core.objects.object_3d import Object3D
from kraken.core.objects.components.component import Component
from kraken.core.objects.attributes.attribute_group import AttributeGroup
from kraken.core.objects.constraints.constraint import Constraint
from kraken.core.objects.operators.operator import Operator


class Traverser(object):
    """Kraken base traverser for any scene item.

    The traverser iterates through all root items and determines the
    dependencies between objects, building an ordered list as it goes. This
    order is then used by the builder to ensure that objects are created and
    evaluated in the correct order. Offset's will then be reliable.

    """

    def __init__(self, name='Traverser'):
        self._rootItems = []
        self.reset()

    # ==================
    # Property Methods
    # ==================
    @property
    def rootItems(self):
        """Gets the root items of this Traverser.

        Returns:
            list: The root items of this Traverser.

        """

        return self._xfo

    def addRootItem(self, item):
        """Adds a new root object to this Traverser

        Args:
            item (SceneItem): The SceneItem to add as a root item.

        Returns:
            bool: True if successful.

        """

        for rootItem in self._rootItems:
            if rootItem.getId() == item.getId():
                return False

        self._rootItems.append(item)

        return True

    def addRootItems(self, items):
        """Adds a bunch of root items to this Traverser

        Args:
            items (SceneItem[]): The SceneItems to add as root items.

        Returns:
            bool: True if successful.

        """

        for item in items:
            self.addRootItem(item)

        return True

    @property
    def items(self):
        """Gets the traversed items of this Traverser.

        Returns:
            list: The traversed items of this Traverser.

        """

        return self._items

    def getItemsOfType(self, typeNames):
        """Gets only the traversed items of a given type.

        Args:
            typeName (str): The name of the type to look for.

        Returns:
            list: The items in the total items list matching the given type.

        """

        if not isinstance(typeNames, list):
            typeNames = [typeNames]

        result = []
        for item in self._items:
            if item.isOfAnyType(typeNames):
                result.append(item)

        return result

    # =============
    # Traverse Methods
    # =============
    def reset(self):
        """Resets all internal structures of this Traverser."""

        self._visited = {}
        self._items = []

    def traverse(self, itemCallback=None, discoverCallback=None,
                 discoveredItemsFirst=True):
        """Visits all objects within this Traverser based on the root items.

        Args:
            itemCallback (func): A callback function to be invoked for each
                item visited.
            discoverCallback (func): A callback to return an array of children
                for each item.

        """

        self.reset()

        if discoverCallback is None:
            discoverCallback = self.discoverBySource

        for item in self._rootItems:
            self.__visitItem(item,
                             itemCallback,
                             discoverCallback,
                             discoveredItemsFirst)

        return self.items

    def __collectVisitedItem(self, item, itemCallback):
        """Doc String.

        Args:
            item (Type): information.
            itemCallback (Type): information.

        """

        if itemCallback is not None:
            itemCallback(item=item, traverser=self)

        self._items.append(item)

    def __visitItem(self, item, itemCallback, discoverCallback, discoveredItemsFirst):
        """Doc String.

        Args:
            item (Type): information.
            itemCallback (Type): information.
            discoverCallback (Type): information.
            discoveredItemsFirst (Type): information.

        Returns:
            Type: information.

        """

        if self._visited.get(item.getId(), False):
            return False

        self._visited[item.getId()] = True

        if hasattr(item, 'getParent') and item.getParent():
            # If this is an attribute and we have not traversed its parent AttributeGroup then skip this
            # and visit the parent so we get this attribute and all others from there (for the sake of attr order)
            if item.isTypeOf("Attribute") and not self._visited.get(item.getParent().getId(), False):
                self._visited[item.getId()] = False
                self.__visitItem(item.getParent(),
                             itemCallback,
                             discoverCallback,
                             discoveredItemsFirst)
                return False

            self.__visitItem(item.getParent(),
                             itemCallback,
                             discoverCallback,
                             discoveredItemsFirst)

        sourcedByConstraintOrOperator = False
        if discoveredItemsFirst:
            for source in item.getSources():
                if isinstance(source, (Constraint, Operator)):
                    sourcedByConstraintOrOperator = True
                    break

        if not discoveredItemsFirst or sourcedByConstraintOrOperator:
            self.__collectVisitedItem(item, itemCallback)

        if discoverCallback:
            if isinstance(item, AttributeGroup):
                discoveredItems = self.discoverChildren(item)
            else:
                discoveredItems = discoverCallback(item)

            if discoveredItems:
                for discoveredItem in discoveredItems:
                    self.__visitItem(discoveredItem,
                                     itemCallback,
                                     discoverCallback,
                                     discoveredItemsFirst)

        if discoveredItemsFirst and not sourcedByConstraintOrOperator:
            self.__collectVisitedItem(item, itemCallback)

        return True

    def discoverChildren(self, item):
        """Doc String.

        Args:
            item (Type): information.

        Returns:
            Type: information.

        """

        result = []

        if isinstance(item, Component):
            subItems = item.getItems().values()
            for subItem in subItems:
                if isinstance(subItem, AttributeGroup):
                    continue
                result.append(subItem)

        elif isinstance(item, Object3D):
            for i in xrange(item.getNumAttributeGroups()):
                result.append(item.getAttributeGroupByIndex(i))
            result += item.getChildren()

        if isinstance(item, AttributeGroup):
            for i in xrange(item.getNumAttributes()):
                result.append(item.getAttributeByIndex(i))

        return result

    def discoverBySource(self, item):
        """Doc String.

        Args:
            item (Type): information.

        Returns:
            Type: information.

        """

        result = []

        for source in item.getSources():

            if isinstance(source, Operator):
                for outputName in source.getOutputNames():
                    operatorOutputs = source.getOutput(outputName)
                    if not isinstance(operatorOutputs, list):
                        operatorOutputs = [operatorOutputs]
                    for operatorOutput in operatorOutputs:
                        if not isinstance(operatorOutput, SceneItem):
                            continue
                        result.append(operatorOutput)

            result.append(source)

        return result
