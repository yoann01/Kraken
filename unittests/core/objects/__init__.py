
import unittest

from attributes import suite as attributeSuite
from components import suite as componentSuite
from constraints import suite as constraintSuite
from operators import suite as operatorSuite

from test_component_group import suite as componentGroupSuite
from test_container import suite as containerSuite
from test_control import suite as controlSuite
from test_ctrlSpace import suite as ctrlSpaceSuite
from test_curve import suite as curveSuite
from test_hierarchy_group import suite as hierarchyGroupSuite
from test_joint import suite as jointSuite
from test_layer import suite as layerSuite
from test_locator import suite as locatorSuite
from test_object_3d import suite as object3DSuite
from test_rig import suite as rigSuite
from test_scene_item import suite as sceneItemSuite
from test_transform import suite as transformSuite

loadAttributeSuite = attributeSuite()
loadComponentSuite = componentSuite()
loadConstraintSuite = constraintSuite()
loadOperatorSuite = operatorSuite()

loadComponentGroupSuite = componentGroupSuite()
loadContainerSuite = containerSuite()
loadControlSuite = controlSuite()
loadCtrlSpaceSuite = ctrlSpaceSuite()
loadCurveSuite = curveSuite()
loadHierarchyGroupSuite = hierarchyGroupSuite()
loadJointSuite = jointSuite()
loadLayerSuite = layerSuite()
loadLocatorSuite = locatorSuite()
loadObject3DSuite = object3DSuite()
loadRigSuite = rigSuite()
loadSceneItemSuite = sceneItemSuite()
loadTransformSuite = transformSuite()


def suite():
    suites = [
        loadAttributeSuite,
        loadComponentSuite,
        loadConstraintSuite,
        loadOperatorSuite,
        loadContainerSuite,
        loadControlSuite,
        loadCtrlSpaceSuite,
        loadCurveSuite,
        loadHierarchyGroupSuite,
        loadJointSuite,
        loadLayerSuite,
        loadLocatorSuite,
        loadObject3DSuite,
        loadRigSuite,
        loadSceneItemSuite,
        loadTransformSuite]

    return unittest.TestSuite(suites)


if __name__ == '__main__':
    runner = unittest.TextTestRunner()
    runner.run(suite())
