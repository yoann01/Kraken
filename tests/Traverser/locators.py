from kraken import plugins
from kraken.core.objects.locator import Locator
from kraken.core.traverser.traverser import Traverser

locA = Locator("locatorA")
locB = Locator("locatorB", parent = locA)
locC = Locator("locatorC", parent = locB)
locD = Locator("locatorD", parent = locB)

trav = Traverser()
trav.addRootItem(locD)
trav.addRootItem(locC)

def callback(**args):
    item = args.get('item', None)
    print 'Visited '+item.getDecoratedPath()

trav.traverse(itemCallback = callback)
