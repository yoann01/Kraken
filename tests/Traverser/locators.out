Failed to find DCC client. Falling back to Python client.
Visited locatorA
Visited locatorA.locatorB
Visited locatorA.locatorB.locatorD
Visited locatorA.locatorB.locatorC
